### PR TITLE
allow fail close webhook admission

### PR DESCRIPTION
**Pull request by deads2k**
Webhook admission needs to allow failing closed.  Even in an alpha state, I don't want to be one DDOS away from having an exposed cluster.

/assign caesarxuchao
/assign sttts

### CODE REVIEW

**Comment by caesarxuchao** [comment]
I think we should make the webook plugin honor the failure policy first: https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/webhook/admission.go#L200.
We shouldn't allow user to specify ineffective value.

**Comment by deads2k** [comment]
> We shouldn't allow user to specify ineffective value.

Wow.  Didn't realize it wasn't even plumbed through.

**Comment by deads2k** [comment]
> I think we should make the webook plugin honor the failure policy first: https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/webhook/admission.go#L200.
> We shouldn't allow user to specify ineffective value.

Done.

File: plugin/pkg/admission/webhook/admission.go [hunk 1]
```diff
@@ -194,7 +194,7 @@ func (a *GenericAdmissionWebhook) Admit(attr admission.Attributes) error {
 			defer wg.Done()
 			if err := a.callHook(ctx, hook, attr); err == nil {
 				return
-			} else if callErr, ok := err.(*ErrCallingWebhook); ok {
+			} else if callErr, ok := err.(*ErrCallingWebhook); (hook.FailurePolicy == nil || *hook.FailurePolicy == v1alpha1.Ignore) && ok {
```
Comment (sttts) [inline plugin/pkg/admission/webhook/admission.go]: as this is pretty critical, better define a bool before the if clauses.

File: plugin/pkg/admission/webhook/admission_test.go [hunk 2]
```diff
@@ -209,6 +211,28 @@ func TestAdmit(t *testing.T) {
...
+				}, {
+					Name:          "internalErr B",
+					ClientConfig:  ccfg,
+					Rules:         matchEverythingRules,
+					FailurePolicy: &policyFail,
+				}, {
+					Name:          "internalErr C",
+					ClientConfig:  ccfg,
+					Rules:         matchEverythingRules,
+					FailurePolicy: &policyFail,
```
Comment (sttts) [inline plugin/pkg/admission/webhook/admission_test.go]: do we have the nil case as well?

**Comment by deads2k** [comment]
comments addressed.

File: plugin/pkg/admission/webhook/admission.go [hunk 3]
```diff
@@ -192,9 +192,12 @@ func (a *GenericAdmissionWebhook) Admit(attr admission.Attributes) error {
 	for i := range hooks {
 		go func(hook *v1alpha1.ExternalAdmissionHook) {
 			defer wg.Done()
+
+			ignoreClientCallFailures := hook.FailurePolicy == nil || *hook.FailurePolicy == v1alpha1.Ignore
+
 			if err := a.callHook(ctx, hook, attr); err == nil {
 				return
-			} else if callErr, ok := err.(*ErrCallingWebhook); ok {
+			} else if callErr, ok := err.(*ErrCallingWebhook); ignoreClientCallFailures && ok {
```
Comment (caesarxuchao) [inline plugin/pkg/admission/webhook/admission.go]: Could you print the `ErrCallingWebhook` separately? Currently the printed warning is prefixed with "rejected by webhook ...", which is misleading.

File: plugin/pkg/admission/webhook/admission_test.go (hunk 2 above)
Comment (caesarxuchao) [reply plugin/pkg/admission/webhook/admission_test.go]: Yeah, above this case there is "match & fail (but allow because fail open on nil)"

**Comment by caesarxuchao** [comment]
One nit and otherwise lgtm.

File: plugin/pkg/admission/webhook/admission.go (hunk 3 above)
Comment (deads2k) [reply plugin/pkg/admission/webhook/admission.go]: > Could you print the ErrCallingWebhook separately? Currently the printed warning is prefixed with "rejected by webhook ...", which is misleading.

updated

File: plugin/pkg/admission/webhook/admission.go [hunk 4]
```diff
@@ -192,16 +192,28 @@ func (a *GenericAdmissionWebhook) Admit(attr admission.Attributes) error {
...
+					// Since we are failing open to begin with, we do not send an error down the channel
+					return
+				}
+
+				glog.Warningf("Failed calling webhook, failing closed %v: %v", hook.Name, err)
 				errCh <- err
+				return
 			}
+
+			glog.Warningf("rejected by webhook %v %t: %v", hook.Name, err, err)
```
Comment (sttts) [inline plugin/pkg/admission/webhook/admission.go]: nit: Rejected

File: plugin/pkg/admission/webhook/admission.go [hunk 5]
```diff
@@ -192,16 +192,28 @@ func (a *GenericAdmissionWebhook) Admit(attr admission.Attributes) error {
...
 				return
-			} else if callErr, ok := err.(*ErrCallingWebhook); ok {
-				glog.Warningf("Failed calling webhook %v: %v", hook.Name, callErr)
-				utilruntime.HandleError(callErr)
-				// Since we are failing open to begin with, we do not send an error down the channel
-			} else {
-				glog.Warningf("rejected by webhook %v %t: %v", hook.Name, err, err)
+			}
+
+			ignoreClientCallFailures := hook.FailurePolicy == nil || *hook.FailurePolicy == v1alpha1.Ignore
```
Comment (liggitt) [inline plugin/pkg/admission/webhook/admission.go]: while in alpha, we should make the default (nil) fail closed. you should have to explicitly request to fail open

File: plugin/pkg/admission/webhook/admission_test.go (hunk 2 above)
Comment (liggitt) [reply plugin/pkg/admission/webhook/admission_test.go]: nil failure policy should default to fail closed. you should have to explicitly request failing open

**Comment by liggitt** [comment]
one comment on default failure policy. fine as a follow-up
/approve no-issue

**Comment by prabushyam** [comment]
@deads2k @caesarxuchao could we cherrypick this onto 1.8?

**Comment by deads2k** [comment]
> @deads2k @caesarxuchao could we cherrypick this onto 1.8?

I don't think so. This effectively new function and the impl in 1.8 was very alpha.  Numerous significant changes are being made in 1.9.

**Pull request merged**